### ISSUE TITLE
Reference images in Sun80 dataset

### ISSUE THREAD

**Issue by SkyeLu** [researchmm/TTSR#19]
Hi, thanks for sharing you great work! You paper said that "Sun80 contains 80 natural images, each paired with several reference images". May I ask that when you test on Sun80 dataset, how do you choose reference images for each input image? I can not tell apart which image is another image's reference image through their names. Besides, since there are several reference images for each image, do you just randomly sample one reference or use all of the references?

**Comment by scutlrr** [comment]
you can refer to this issue, Sun80 is similar to CUFED5.
https://github.com/researchmm/TTSR/issues/3

**Comment by FuzhiYang** [comment]
In Sun80 dataset, there are 80 test images, each accompanied with about 20 reference images. We just randomly choose one as ref to test. The whole performance will not variant too much.

**Comment by scutlrr** [comment]
According to the author, I stitched the five reference pictures in CUFED5 for testing, but the test result was only 26.8+db. Have you tested it on CUFED5?

**Comment by scutlrr** [comment]
I used the following code to stitch five reference pictures in CUFED5 and then used it (2500x500) as a reference picture for testing, but the test result was only 26.83db (27.09db in your paper). Can you provide some suggestions?

`import glob
import numpy as np
import os
from PIL import Image
from imageio import imread, imsave

if __name__ == '__main__':
    ref_lv1 = sorted(glob.glob(os.path.join('./ref_test', '*_1.png')))
    ref_lv2 = sorted(glob.glob(os.path.join('./ref_test', '*_2.png')))
    ref_lv3 = sorted(glob.glob(os.path.join('./ref_test', '*_3.png')))
    # ref_lv4 = sorted(glob.glob(os.path.join('./ref_test', '*_4.png')))
    # ref_lv5 = sorted(glob.glob(os.path.join('./ref_test', '*_5.png')))
    save_path = './stitch_top3_image'

    # for ref_1, ref_2, ref_3, ref_4, ref_5 in zip(ref_lv1, ref_lv2, ref_lv3, ref_lv4, ref_lv5):
    for ref_1, ref_2, ref_3 in zip(ref_lv1, ref_lv2, ref_lv3):
        lv1 = imread(ref_1)
        lv1 = np.array(Image.fromarray(lv1).resize((500, 500), Image.BICUBIC))
        lv2 = imread(ref_2)
        lv2 = np.array(Image.fromarray(lv2).resize((500, 500), Image.BICUBIC))
        lv3 = imread(ref_3)
        lv3 = np.array(Image.fromarray(lv3).resize((500, 500), Image.BICUBIC))
        lv4 = imread(ref_4)
        lv4 = np.array(Image.fromarray(lv4).resize((500, 500), Image.BICUBIC))
        lv5 = imread(ref_5)
        lv5 = np.array(Image.fromarray(lv5).resize((500, 500), Image.BICUBIC))
        Stitch = np.zeros((1500, 500, 3))
        Stitch[:500, :500, :] = lv1
        Stitch[500:1000, :500, :] = lv2
        Stitch[1000:1500, :500, :] = lv3
        Stitch[1500:2000, :500, :] = lv4
        Stitch[2000:2500, :500, :] = lv5
        imsave(os.path.join(save_path, os.path.basename(ref_1)), Stitch.astype(np.uint8))
`

**Comment by FuzhiYang** [comment]
You do not need to bicubicly interpolate each ref to 500x500. You can just zero padding each ref to 500x500. The bicubic interpolation will demage the high-res textures severely.

**Comment by scutlrr** [comment]
> You do not need to bicubicly interpolate each ref to 500x500. You can just zero padding each ref to 500x500. The bicubic interpolation will demage the high-res textures severely.

Well, I will try it again. Thank you very much!

**Comment by SkyeLu** [comment]
> you can refer to this issue, Sun80 is similar to CUFED5.
> #3

Thanks!

**Comment by SkyeLu** [comment]
> In Sun80 dataset, there are 80 test images, each accompanied with about 20 reference images. We just randomly choose one as ref to test. The whole performance will not variant too much.

Got it. Thanks a lot!

**Comment by scutlrr** [comment]
> You do not need to bicubicly interpolate each ref to 500x500. You can just zero padding each ref to 500x500. The bicubic interpolation will demage the high-res textures severely.

I have re-stitched the pictures and tested it with a pre-trained model (TTSR-rec.pt), the result (27.086db) is very close to the result in your paper, thank you again!

**Comment by flybiubiu** [comment]
> > 您无需将每个参照三次三次插入到500x500。您可以将每个参考零填充到500x500。双三次插值将严重破坏高分辨率纹理。
> 
> 我重新拼接了图片，并使用预先训练的模型（TTSR-rec.pt）对其进行了测试，结果（27.086db）非常接近您的论文结果，再次感谢您！

Thanx!Can you provide the details of the splicing？

**Comment by flybiubiu** [comment]
> > You do not need to bicubicly interpolate each ref to 500x500. You can just zero padding each ref to 500x500. The bicubic interpolation will demage the high-res textures severely.
> 
> I have re-stitched the pictures and tested it with a pre-trained model (TTSR-rec.pt), the result (27.086db) is very close to the result in your paper, thank you again!

Thankx！Can you provide the details of the splicing？Please be more specific.I always have some error.

**Comment by scutlrr** [comment]
img=np.zeros((2500,500,3))
Read the ref image and assign it to the corresponding position(in img)
h, w = ref_1.shape[:2]
For example, img[0:h,0:w,:]=ref_1, and so on

**Comment by scutlrr** [comment]
> Hi, thanks for sharing you great work! You paper said that "Sun80 contains 80 natural images, each paired with several reference images". May I ask that when you test on Sun80 dataset, how do you choose reference images for each input image? I can not tell apart which image is another image's reference image through their names. Besides, since there are several reference images for each image, do you just randomly sample one reference or use all of the references?

hi, can you share the link to the Sun80 dataset? The dataset I downloaded from the Internet has only 7 refs instead of 20 as the author said.

**Comment by flybiubiu** [comment]
> > Hi, thanks for sharing you great work! You paper said that "Sun80 contains 80 natural images, each paired with several reference images". May I ask that when you test on Sun80 dataset, how do you choose reference images for each input image? I can not tell apart which image is another image's reference image through their names. Besides, since there are several reference images for each image, do you just randomly sample one reference or use all of the references?
> 
> hi, can you share the link to the Sun80 dataset? The dataset I downloaded from the Internet has only 7 refs instead of 20 as the author said.

7 refs ,too.
ps:
Thankx!

**Comment by flybiubiu** [comment]
> > Hi, thanks for sharing you great work! You paper said that "Sun80 contains 80 natural images, each paired with several reference images". May I ask that when you test on Sun80 dataset, how do you choose reference images for each input image? I can not tell apart which image is another image's reference image through their names. Besides, since there are several reference images for each image, do you just randomly sample one reference or use all of the references?
> 
> hi, can you share the link to the Sun80 dataset? The dataset I downloaded from the Internet has only 7 refs instead of 20 as the author said.



> > You do not need to bicubicly interpolate each ref to 500x500. You can just zero padding each ref to 500x500. The bicubic interpolation will demage the high-res textures severely.
> 
> I have re-stitched the pictures and tested it with a pre-trained model (TTSR-rec.pt), the result (27.086db) is very close to the result in your paper, thank you again!

ref1 = []
                ref2 = []
                ref3 = []
                ref4 = []
                ref5 = []
                for i_batch, sample_batched in enumerate(self.dataloader['test']['1']):
                    ref1.append(sample_batched['Ref'])
                for i_batch, sample_batched in enumerate(self.dataloader['test']['2']):
                    ref2.append(sample_batched['Ref'])
                for i_batch, sample_batched in enumerate(self.dataloader['test']['3']):
                    ref3.append(sample_batched['Ref'])
                for i_batch, sample_batched in enumerate(self.dataloader['test']['4']):
                    ref4.append(sample_batched['Ref'])
                for i_batch, sample_batched in enumerate(self.dataloader['test']['5']):
                    ref5.append(sample_batched['Ref'])
                Ref = []
                for i in range(126):
                    img = np.zeros((1,3, 2500, 500))
                    h1, w1 = ref1[i].shape[2:]
                    #print(ref1[i].shape)
                    h2, w2 = ref2[i].shape[2:]
                    h3, w3 = ref3[i].shape[2:]
                    h4, w4 = ref4[i].shape[2:]
                    h5, w5 = ref5[i].shape[2:]
                    img[:,:,0:h1, 0:w1] = ref1[i]
                    img[:,:,500:h2+500, 0:w2] = ref2[i]
                    img[:,:,1000:h3+1000, 0:w3] = ref3[i]
                    img[:,:,1500:h4+1500, 0:w4] = ref4[i]
                    img[:,:,2000:h5+2000, 0:w5] = ref5[i]
                    img = torch.from_numpy(img)
                    Ref.append(img)

                    #self.dataloader['test']['1']

                for i_batch, sample_batched in enumerate(self.dataloader['test']['1']):
                    cnt += 1
                    sample_batched = self.prepare(sample_batched)
                    lr = sample_batched['LR']
                    lr_sr = sample_batched['LR_sr']
                    hr = sample_batched['HR']
                    sample_batched['Ref'] = Ref[i_batch].float().to(self.device)
                    ref = sample_batched['Ref']
                    ref_sr = sample_batched['Ref_sr']

<img width="489" alt="Screen Shot 2020-10-28 at 4 45 02 PM" src="https://user-images.githubusercontent.com/20927755/97412767-f4c1b600-193c-11eb-9757-251f330532ea.png">
Thanx! What wrong with it?
ps:
Can I exchange a chat tool with you?hhhhhh

**Comment by sfahad19** [comment]
Hi, could you please kindly share the link of the Sun80 dataset?
